### PR TITLE
fix: insufficient permission for ingestion server when missing required service roles for ecs and auto scaling

### LINKED ISSUE #226: fail to deploy ingestion and data processing stacks with insufficient permission
## Summary

Deploy a pipeline in a new AWS account without creating any computing resource before.

## Steps to reproduce

1. create a project
2. configure the pipeline using KDS as sink

## What is the current bug behavior?

fail to create ingestion stack with below error

> 2023-09-15 22:25:52 UTC+0800	IngestionServerclickstreamingestionserviceecscapacityprovider8FF654A6	
CREATE_FAILED
Resource handler returned message: "Invalid request provided: CreateCapacityProvider error: The ECS Service Linked Role does not exist. Please ensure that it exists and try again. (Service: AmazonECS; Status Code: 400; Error Code: ClientException; Request ID: 70c22788-82a4-407b-9a47-01f782820689; Proxy: null)" (RequestToken: 507dd7fb-e9cd-ed70-de6a-6af1cf6998eb, HandlerErrorCode: InvalidRequest) when creating ingestion stack

> 2023-09-15 22:20:53 UTC+0800 | NestedStackCreateEMRServelsssApplicationCustomResourceB1D0767E | CREATE_FAILED | Received response status [FAILED] from custom resource. Message returned: Access denied when calling CreateServiceLinkedRole operation Logs: /aws/lambda/Clickstream-DataProcessin-NestedStackCreateEMRServ-1mD9JBVEmCwN at de_ValidationExceptionRes (/var/task/index.js:20305:25) at de_CreateApplicationCommandError (/var/task/index.js:19731:23) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async /var/task/index.js:6470:24 at async /var/task/index.js:3878:22 at async /var/task/index.js:7947:42 at async /var/task/index.js:1337:26 at async createEMRServerlessApp (/var/task/index.js:43788:20) at async _handler (/var/task/index.js:43760:27) at async Runtime.handler (/var/task/index.js:43743:17) (RequestId: 21c4bb08-77f1-4e57-b4f3-e536ae9e90b4) in data processing stack

> ~~API: autoscaling:CreateAutoScalingGroup The default Service-Linked Role for Auto Scaling could not be created. com.amazonaws.services.identitymanagement.model.AmazonIdentityManagementException: User: arn:aws:sts::1234567890:assumed-role/gcr-solutions-clickstream-ClickStreamApiStackActio-1K5K4UTZSFHVA/gcr-solutions-clickstream-ClickStreamApiStackActio-l2PreoxogsBW is not authorized to perform: iam:CreateServiceLinkedRole on resource: arn:aws:iam::1234567890:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling because no identity-based policy allows the iam:CreateServiceLinkedRole action (Service: AmazonIdentityManagement; Status Code: 403; Error Code: AccessDenied; Request ID: d8ec6f35-1829-4768-8ab1-bfcfbf254e9e; Proxy: null)~~ resolved by #228 

## Possible fixes

In #115 we introduced a custom resource to manage the lifecycle of EMR Serverless application instead of native CloudFormation resource type. The function will fail to create the emr serverless application with the missing permission to create service link role if the role does not exist yet. Add below policy to the Lambda execution role should fix it,

```json
        {
            "Action": [
                "iam:CreateServiceLinkedRole"
            ],
            "Resource": [
"arn:aws:iam::123456789012:role/aws-service-role/ops.emr-serverless.amazonaws.com/AWSServiceRoleForAmazonEMRServerless"
            ],
            "Effect": "Allow"
        },
```

For the insufficient permission to create ingestion server, the stack was launched by the Step Function workflow. We need grant below permissions to the function which launches the stack with the permissions for creating missing service link roles,

```
        {
            "Action": [
                "iam:PassRole",
                "iam:CreateServiceLinkedRole"
            ],
            "Resource": [
                "arn:aws:iam::123456789012:role/aws-service-role/ecs.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_ECSService",
                "arn:aws:iam::123456789012:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling",
                "arn:aws:iam::123456789012:role/aws-service-role/ecs.amazonaws.com/AWSServiceRoleForECS",
                "arn:aws:iam::123456789012:role/aws-service-role/elasticloadbalancing.amazonaws.com/AWSServiceRoleForElasticLoadBalancing"
            ],
            "Effect": "Allow"
        },
```

---

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

close #226 

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/stack-action-state-machine-construct.ts [hunk 1]
```diff
@@ -123,6 +123,9 @@ export class StackActionStateMachine extends Construct {
             `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:instance-profile/Clickstream*`,
             `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:role/aws-service-role/ecs.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_ECSService`,
             `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling`,
+            `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:role/aws-service-role/ecs.amazonaws.com/AWSServiceRoleForECS`,
```
Comment (zxkane) [inline src/control-plane/backend/stack-action-state-machine-construct.ts:126]: Let's move the service link role resources to sperated statement like the issue proposed,

```
{
            "Action": [
                "iam:PassRole",
                "iam:CreateServiceLinkedRole"
            ],
            "Resource": [
                "arn:aws:iam::123456789012:role/aws-service-role/ecs.application-autoscaling.amazonaws.com/AWSServiceRoleForApplicationAutoScaling_ECSService",
                "arn:aws:iam::123456789012:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling",
                "arn:aws:iam::123456789012:role/aws-service-role/ecs.amazonaws.com/AWSServiceRoleForECS",
                "arn:aws:iam::123456789012:role/aws-service-role/elasticloadbalancing.amazonaws.com/AWSServiceRoleForElasticLoadBalancing"
            ],
            "Effect": "Allow"
        },
```

File: src/control-plane/backend/stack-action-state-machine-construct.ts [hunk 2]
```diff
@@ -123,6 +123,9 @@ export class StackActionStateMachine extends Construct {
...
+            `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:role/aws-service-role/elasticloadbalancing.amazonaws.com/AWSServiceRoleForElasticLoadBalancing`,
+            `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:role/aws-service-role/ops.emr-serverless.amazonaws.com/AWSServiceRoleForAmazonEMRServerless`,
```
Comment (zxkane) [inline src/control-plane/backend/stack-action-state-machine-construct.ts:128]: It's not necessary for backend API. The EMR serverless application is managed in a sperated Lambda, which uses the execution role used by the Lambda function.

Comment (zxkane) [reply src/control-plane/backend/stack-action-state-machine-construct.ts:128]: @amliuyong Let's fix it in the role of the function to manage the emr serverless application.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/stack-action-state-machine-construct.ts (hunk 2 above)
Comment (zxkane) [reply src/control-plane/backend/stack-action-state-machine-construct.ts:128]: emr permission missing fixed in #253

**Comment by tyyzqmf** [comment]
fixed in https://github.com/awslabs/clickstream-analytics-on-aws/pull/253

**Comment by zxkane** [comment]
#253 only resolved the insufficient permissions for creating data processing stack, not **ingestion server** stack.

File: src/control-plane/backend/stack-action-state-machine-construct.ts (hunk 2 above)
Comment (tyyzqmf) [reply src/control-plane/backend/stack-action-state-machine-construct.ts]: done.

File: src/control-plane/backend/stack-action-state-machine-construct.ts (hunk 1 above)
Comment (tyyzqmf) [reply src/control-plane/backend/stack-action-state-machine-construct.ts:133]: done.

**Review by zxkane — APPROVED** [review]

**Pull request merged**